src/modules/job-manager/restart.c

```diff
@@ -138,9 +138,25 @@ static int restart_map_cb (struct job *job, void *arg)
         return -1;
     if ((job->flags & FLUX_JOB_WAITABLE))
         wait_notify_active (ctx->wait, job);
-    if (event_job_action (ctx->event, job) < 0) {
-        flux_log_error (ctx->h, "%s: event_job_action id=%ju",
-                        __FUNCTION__, (uintmax_t)job->id);
+    if (job->state == FLUX_JOB_STATE_SCHED) {
+       /* The flux-restart event is posted to jobs in SCHED state to
+        * transition back to PRIORITY so that the queue priority can
+        * be re-established.
+        */
+        if (event_job_post_pack (ctx->event,
```

Comment:
Why is this a special case in the restart map function?  Wouldn't this post the `flux-restart` event for all jobs as they pass through SCHED even if that's not their final state?  (Hmm, just did a test where job-manager is reloaded and did not see that so I must not understand).

It would _seem_ like the straightforward way to implement this would be to iterate over jobs _after_ they have all been replayed, and post the restart event for each job found to be in SCHED state.  Then handle this event like any other in `event_job_update()`, and in `event_job_action()`, when handing PRIORITY state,  dequeue job from the alloc queue (if enqueued).

Reply:
Oh, I just realized the map function is mapping over jobs not events.  Duh.  Sorry.

Reply:
It would still be nice if this event was present in the state machine logic and not treated as magical during restart.

Reply:
> Why is this a special case in the restart map function? Wouldn't this post the flux-restart event for all jobs as they pass through SCHED even if that's not their final state? (Hmm, just did a test where job-manager is reloaded and did not see that so I must not understand).

so from how i was reading the code

```
    if (!(job = job_create_from_eventlog (id, eventlog)))
        goto done;
    if (cb (job, arg) < 0)
        goto done;
```

the callback is called once after the entire eventlog has been processed.

> It would seem like the straightforward way to implement this would be to iterate over jobs after they have all been replayed, and post the restart event for each job found to be in SCHED state. Then handle this event like any other in event_job_update(), and in event_job_action(), when handing PRIORITY state, dequeue job from the alloc queue (if enqueued).

The reason I did it this was basically to avoid the call to `event_job_action()` after `job_create_from_eventlog()` was complete.  Logically what the code (before this PR) is doing on restart is:

```
for each eventlog
    for each event in the eventlog
         event_job_update ();
    event_job_action ();
```

the event job action at the end would place the job into the alloc queue.  In effect, what I'm doing is:

```
for each eventlog
    for each event in the eventlog
         event_job_update ();
    if (job in SCHED)
        event_job_post_pack ("flux-restart");
    else
       event_job_action ();
```

so it never call `event_job_action()` and the job never enters the alloc queue if it is in state SCHED.  I think you're proposing:


```
for each eventlog
    for each event in the eventlog
         event_job_update ();
    event_job_action ();
if (job in SCHED)
    event_job_post_pack ("flux-restart");
```

and then `event_job_update ()`  would have to dequeue if needed?

Is it in your opinion this is less confusing / more straightforward?  I think this would be fine as well.  Or perhaps some comments are needed to describe whats going on in the code I've pushed.

Reply:
Apologies - my original misread took me off in the weeds!

It still seems like it would be be more straightforward if the restart code recreated the state based on the eventlogs (enqueuing SCHED jobs) then on the way out posted `flux-restart` to jobs in SCHED state?  E.g. inside the zhashx iterator in `restart_from_kvs()`?  Then `event_job_action()` PRIORITY state entry would need to dequeue the job if `job->alloc_queued` is true.